### ISSUE TITLE
Allow the use of user-provided history class

### ISSUE THREAD

**Issue by ehuelsmann** [jonasbn/perl-workflow#106]
# Issue Template

### Description

The library currently does not support using a different history class than the one that comes with Workflow. If a user of the library has additional fields to serialize and restore, the current setup does not suffice for that.

In my case: the users of my application have an "id" I want to link them to (next to putting their user names in the "username" field -- yet there's no way to do so except using a database trigger. But even when using a database trigger, there's no way to restore the data into a history object.

# Bugs and Issues

### Specify what version was used

Any version that's currently released (current is 1.53).

**Comment by ehuelsmann** [comment]
In order to be able to use configurable `Workflow::History` classes, there's a design problem in `Workflow::Persister->fetch_history()`: its definition states that it is to return `Workflow::History` instances, which tightly couples the persister with history. The integration between the persister and `Workflow` is much lighter: the persister is documented to return a hash with initialization parameters.

I'd propose to change the persister to be required to return an array of hashes (or arrays) with initialization arguments for the configurable `Workflow::History` class.

**Comment by ehuelsmann** [comment]
There's an important question in this issue that we need to solve before I can develop this change: Do we make it the responsibility of the factory to instantiate history objects for the workflow? Or is that the responsibility of the workflow itself?

And in case of the latter, do we make it the responsibility of the Workflow class to make sure the history class/module is loaded? Or is that a responsibility of the factory?

Note that resolution of this issue depends on #124 .

**Comment by ehuelsmann** [comment]
#138 implements a change where the answer in the above comment is implemented as:

```
The workflow is responsible for the instantiation of (its) history.
```

Since the workflow is also responsible for generating history, with the configured history class, it's an automatic choice to allocate this responsibility to the workflow. The Factory currently has no role in history handling; the only thing it does is separating the concerns between the workflow and the persister.

@jonasbn I suggest to close this issue on the grounds that the merge of #138 implemented #124 and the additional reasoning in this comment.

**Comment by jonasbn** [comment]
We close this is, as we consider it as addressed by #138

**Comment by ehuelsmann** [comment]
I was too early to propose to close this. I'm working on a PR which implements the remainder.
